### ISSUE TITLE
Mixed types aren't evaluated separately and raise incorrect errors on API-compatible attribute access

### ISSUE THREAD

**Issue by tigerhawkvok** [astral-sh/ty#4395]
### Summary

Sample:

```py
from collections.abc import Sequence
from typing import Optional, TypeVar, Union, overload
import numpy as np
from pandas import DataFrame as pdFrame, Series as pdSeries
from pyspark.pandas import DataFrame as pspdFrame, Series as pspdSeries

PandasLike = pdFrame | pspdFrame

# Incoming may be one, the other, or an unknown mix of DataFrame types
PSPD_FRAME = TypeVar("PSPD_FRAME", pdFrame, pspdFrame, PandasLike)

@overload
def pspdLoc(_df:PSPD_FRAME, _sliceRows:Union[slice, np.ndarray, None], _sliceCols:Optional[Sequence[str]]= None) -> PSPD_FRAME: ...
@overload
def pspdLoc(_df:PSPD_FRAME, _sliceRows:Union[slice, np.ndarray, None], _sliceCols:Union[slice, Sequence[str], str, None]= None) -> Union[PSPD_FRAME, pspdSeries, pdSeries]: ...
def pspdLoc(_df:PSPD_FRAME, _sliceRows:Union[slice, np.ndarray, None], _sliceCols:Union[slice, Sequence[str], str, None]= None):
    """
    A wrapper for `.loc` that works around type hint
    idiosyncrasies for pandas-on-spark
    """
    if _sliceRows is None:
        _sliceRows = slice(None)
    if _sliceCols is None:
        # Incorrectly not evaluating each constraint separately, reporting:
        #
        # Invalid access to descriptor attribute `loc` on type `PSPD_FRAME@pspdLoc`: Expected `IndexingMixin`, found `pandas.core.frame.DataFrame | pyspark.pandas.frame.DataFrame[Unknown]`
        # info: Argument to function `IndexingMixin.loc` is incorrect
        # info: This access implicitly calls `__get__` on a descriptor of type `property`
        # info: element `pyspark.pandas.frame.DataFrame[Unknown]` of union `pandas.core.frame.DataFrame | pyspark.pandas.frame.DataFrame[Unknown]` is not assignable to `IndexingMixin`
        #
        # Both of these dataframes expose the `.loc` accessor -- 
        # by _design_, since Pandas-on-Spark is intentionally API compatible with Pandas
        _sliced = _df.loc[_sliceRows]
    else:
        _sliced = _df.loc[_sliceRows, _sliceCols]
    assert _sliced is not None
    return _sliced
```

The error disappears if the unioned type `PandasLike` is removed from the `TypeVar`, which makes the `TypeVar` not correctly describe the function while not actually changing any types flowing through it. 

### Version

0.0.74

**Comment by carljm** [comment]
Thanks for the report! This is a valid bug; it's #1503 and on our roadmap to fix soon.
